File: examples/factorial.clj
(ns clojure.examples.factorial)

(defn fact [x]
    (loop [n x f 1]
        (if (= n 1)
            f
            (recur (dec n) (* f n)))))

(defn test [times]
    (loop [rem times]
        (if (> rem 0)
            (do (fact 20)
                (recur (dec rem))))))

;; on my machine
;; clojure jvm : 25 sec
;; clojure pypy : 48 sec

(time (test 1999999))

File: examples/factorial.py
def fact(x):
    f = 1
    n = x
    while True:
        if n == 1:
            return f
        else:
            f = f * n
            n = n - 1
            continue


def test(times):
    rem = times
    while True:
        if rem > 0:
            fact(20000)
            rem = rem - 1
            continue
        else:
            return


test(20)
